src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.CopyFile.OSX.cs

```diff
@@ -0,0 +1,111 @@
+
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using Microsoft.Win32.SafeHandles;
+
+namespace System.IO
+{
+    internal static partial class FileSystem
+    {
+        public static partial void CopyFile(string sourceFullPath, string destFullPath, bool overwrite)
+        {
+            // Fail fast for blatantly copying onto self
+            if (sourceFullPath == destFullPath)
+            {
+                if (overwrite)
+                {
+                    throw new IOException(SR.Format(SR.IO_SharingViolation_File, destFullPath));
+                }
+
+                throw new IOException(SR.Format(SR.IO_FileExists_Name, destFullPath));
+            }
+
+            // Start by locking, creating relevant file handles, and reading out file status info.
+            using SafeFileHandle src = SafeFileHandle.OpenReadOnly(sourceFullPath, FileOptions.None, out var fileStatus);
+            UnixFileMode filePermissions = SafeFileHandle.GetFileMode(fileStatus);
+
+            // Read FileStatus of destination file to determine how to continue
+            int destError = Interop.Sys.Stat(destFullPath, out var destStat);
+
+            // Interpret the error from stat
+            if (destError != 0)
+            {
+                // Some error, let's see what it is:
+                Interop.ErrorInfo error = Interop.Sys.GetLastErrorInfo();
+
+                // Destination not existing is expected, so if this is the case we try clonefile,
+                // otherwise we got some other error that the fallback code can deal with.
+                if (error.Error == Interop.Error.ENOENT)
+                {
+                    goto tryCloneFile;
+                }
+                else
+                {
+                    goto tryFallback;
+                }
+            }
+            else
+            {
+                // Destination exists:
+                if (!overwrite)
+                {
+                    // Throw an error if we're not overriding
+                    throw Interop.GetExceptionForIoErrno(new Interop.ErrorInfo(Interop.Error.EEXIST));
+                }
+                if (fileStatus.Dev != destStat.Dev)
+                {
+                    // On different device
+                    goto tryFallback;
+                }
+                else if (fileStatus.Ino == destStat.Ino)
+                {
+                    // Copying onto itself
+                    throw new IOException(SR.Format(SR.IO_SharingViolation_File, destFullPath));
+                }
+                else
+                {
+                    goto tryDelete;
+                }
+            }
+
+            // Try deleting destination:
+            tryDelete:
+            {
+                // Delete the destination. This should fail on directories. And update the mode.
+                // Get a lock to the dest file for compat reasons, and then delete it.
```

Comment:
"for compat reasons"... what are those reasons?

Reply:
I'm pretty sure it's that `Copy` is not supposed to write over a file that is locked. Will confirm this and then update the comment to be clearer